### ISSUE TITLE
[fltk] FLTK 1.3 is still needed as 1.4 is not fully backwards compatible

### ISSUE THREAD

**Issue by CendioOssman** [msys2/MINGW-packages#22769]
### Description / Steps to reproduce the issue

1. Install FLTK from MSYS
2. Download TigerVNC source
3. Build TigerVNC

### Expected behavior

TigerVNC builds

### Actual behavior

TigerVNC fails to build (and if it would build, it would not behave correctly).

### Verification

- [X] I have verified that my MSYS2 is up-to-date before submitting the report (see https://www.msys2.org/docs/updating/)

### Windows Version

10.0.20348

### MINGW environments affected

- [X] MINGW64
- [ ] MINGW32
- [ ] UCRT64
- [ ] CLANG64
- [ ] CLANGARM64

**Comment by CendioOssman** [comment]
Related, FLTK 1.3.10 was released at the same time as FLTK 1.4.0 and contains new functionality that TigerVNC wants. So it would be nice to get that new version packaged in to MSYS.

**Comment by CendioOssman** [comment]
As a temporary workaround, we've put this in place: https://github.com/TigerVNC/tigervnc/commit/f06e574be1e6aa969084ca3ff4918dbc2ee32f2b

**Comment by lazka** [comment]
According to upstream (https://www.fltk.org/roadmap.php) "...FLTK 1.3 is no longer in active development...".

We could add back a 1.3 version if really needed. How large are the incompatibilities? All our reverse deps built just fine with it.

**Comment by CendioOssman** [comment]
> We could add back a 1.3 version if really needed. How large are the incompatibilities? All our reverse deps built just fine with it.

We have not dug in to it. TigerVNC is a very advanced user of FLTK, and FLTK 1.4 has been in development for many years. So there are likely many things that need to be adjusted. As an example, we've got a report that the rendering pipeline is not compatible any more (linked above).

To make things more complicated, we will likely need to support FLTK 1.3 for some time, as I would not assume 1.4 will be available in older Linux distributions.

In short, FLTK 1.4 support in TigerVNC is not something anyone will have time to look at anytime soon.

As a comparison, macOS homebrew now packages both 1.4 and 1.3: https://formulae.brew.sh/formula/fltk@1.3

I'm fairly sure this kind of overlap was also done for some time when 1.3 was released, which also wasn't fully backwards compatible with the previous version of FLTK.

**Comment by lazka** [comment]
> There would be file conflict if both fltk 1.3 and 1.4 is installed. I think the underlying issue should be fixed in tigervnc.

There are two way to go about it:

* just conflict and restrict what can be installed
* split 1.3 into a runtime and a dev package, and only the dev package conflicts with fltk 1.4

**Comment by CendioOssman** [comment]
Conflicts are likely fine for us for the foreseeable future. We usually statically link stuff for distribution, so FLTK 1.3 just needs to be installed when building.

**Comment by felixaszx** [comment]
I will be trying to add back 1.3 very soon if noone is working on it now.

**Comment by felixaszx** [comment]
@lazka Apparently for our own distribution (MSYS2) of TigerVNC, we are linking fltk dynamically. So if we decided to go with conflicting packages then a bunch of other packages will be affected.

**Comment by felixaszx** [comment]
@CendioOssman Not yet, but started now. Like @lazka said, I think we can do conflict packages to solve this issue.

**Comment by Biswa96** [comment]
I have added fltk 1.3 as a separate package which is conflicts with fltk 1.4. Could you test the packages from CI artifacts please? Link https://github.com/msys2/MINGW-packages/actions/runs/14017340243

**Comment by CendioOssman** [comment]
They seem to work nicely! Thanks!

Note that there have been two releases of FLTK 1.3, so 1.3.11 is the current release on that branch.

**Comment by Biswa96** [comment]
I have updated the version to 1.3.11.

**Comment by Biswa96** [comment]
fltk 1.3 binary package is available in mingw repository https://packages.msys2.org/base/mingw-w64-fltk1.3 now. Please feel free to ask any question. I have a local VNC and RDP setup to test tigervnc issues.
